src/sonic-dhcp-server/dhcp_server/dhcprelayd.py

```diff
@@ -0,0 +1,197 @@
+
+    def _start_dhcrelay_process(self, new_dhcp_interfaces, dhcp_server_ip):
+        # To check whether need to kill dhcrelay process
+        kill_res = self._kill_exist_relay_releated_process(new_dhcp_interfaces, "dhcrelay")
+        if kill_res == NOT_KILLED:
+            # Means old running status consistent with the new situation, no need to run new
+            return
+
+        # No need to start new dhcrelay process
+        if len(new_dhcp_interfaces) == 0:
+            return
+
+        cmds = ["/usr/sbin/dhcrelay", "-d", "-m", "discard", "-a", "%h:%p", "%P", "--name-alias-map-file",
+                "/tmp/port-name-alias-map.txt"]
+        for dhcp_interface in new_dhcp_interfaces:
+            cmds += ["-id", dhcp_interface]
+        cmds += ["-iu", "docker0", dhcp_server_ip]
+        subprocess.Popen(cmds)
+
+    def _start_dhcpmon_process(self, new_dhcp_interfaces):
+        # To check whether need to kill dhcrelay process
+        kill_res = self._kill_exist_relay_releated_process(new_dhcp_interfaces, "dhcpmon")
+        if kill_res == NOT_KILLED:
+            # Means old running status consistent with the new situation, no need to run new
+            return
+
+        # No need to start new dhcrelay process
+        if len(new_dhcp_interfaces) == 0:
+            return
+
+        for dhcp_interface in new_dhcp_interfaces:
+            cmds = ["/usr/sbin/dhcpmon", "-id", dhcp_interface, "-iu", "docker0", "-im", "eth0"]
+            subprocess.Popen(cmds)
+
+    def _kill_exist_relay_releated_process(self, new_dhcp_interfaces, process_name):
+        old_dhcp_interfaces = set()
+        target_proc = None
+
+        # Get old dhcrelay process and get old dhcp interfaces
+        for proc in psutil.process_iter():
+            if proc.name() == process_name:
+                cmds = proc.cmdline()
+                index = 0
+                target_proc = proc
+                while index < len(cmds):
+                    if cmds[index] == "-id":
+                        old_dhcp_interfaces.add(cmds[index + 1])
+                        index += 2
+                    else:
+                        index += 1
+        if target_proc is None:
+            return NOT_FOUND_PROC
+
+        # No need to kill
+        if process_name == "dhcrelay" and old_dhcp_interfaces == new_dhcp_interfaces or \
+           process_name == "dhcpmon" and old_dhcp_interfaces == (new_dhcp_interfaces):
+            return NOT_KILLED
+
+        target_proc.terminate()
+        target_proc.wait()
+        return KILLED_OLD
+
+    def _get_dhcp_server_ip(self):
+        dhcp_server_ip_table = swsscommon.Table(self.db_connector.state_db, DHCP_SERVER_IPV4_SERVER_IP)
+        while True:
+            state, ip = dhcp_server_ip_table.hget(DHCP_SERVER_INTERFACE, "ip")
+            if state:
+                return ip
+            else:
+                syslog.syslog(syslog.LOG_WARNING, "Cannot get dhcp server ip")
+                time.sleep(2)
```

Comment:
we always expect dhcprelayd start after dhcp server registered its IP in state_db
-> Retry 10 times, sleep 10s, Log Info
-> Fail with Err log

Reply:
Fixed